rfcs/2024-05-21-ui-component-model.md

```diff
@@ -0,0 +1,248 @@
+# UI component model
+
+Authors: Gordon Brander
+
+## Overview
+
+Context: converge on a default UI component model for LLM-generated UI.
+
+In a FE context, “UI component” means a bundle of UI and behavior that together describe an island of interactivity. There are various popular component models, including React, WebComponents, Vue, etc. In keeping with common FE terminology, this rfc uses the term “component” to mean “bundle of code that describes an island of UI”. For our specific context, a UI component is probably a type of module.
+
+## Goals
+
+### Product goals
+
+- Easy for LLMs to generate
+    - Leverages patterns and/or frameworks that are widely present in the training data, or can be learned within a small context window.
+- Easy for humans to edit
+    - [Maybe it’s not even code?](https://x.com/threepointone/status/1792930000766677034) Or maybe it’s a hybrid of code and plain language?
+- Leverages familiar or established patterns for UI development
+- Conformable with existing web FE toolchains.
+
+### Technical goals
+
+- Components are **encapsulated** (P1)
+    - A component may control its child tree, but not its siblings or parent
+- Components have **inputs**, and **output UI** and **events** (P1)
+    - Components can be understood as pure-ish functions that receive props and return a view description (we may allow cheats for local component state ala hooks)
+    - Components are **black boxes**
+    - Components are decoupled and only communicate via input and output channels.
+- Component **inputs** and **outputs** are **statically-typed** (P1)
+    - E.g. via TypeScript
+    - Allows the runtime to enforce data policies on component
+- Components have **local state** (P1)
+    - State is encapsulated within component
+    - Components may pass state down to child components as input
+        - E.g. the React **[“lifting state up” pattern](https://legacy.reactjs.org/docs/lifting-state-up.html)**.
+    - Local state may be **persisted**.
```

Comment:
Can we do just via outputs and inputs? The "pass the state to yourself" pattern? (DX might hide that of course). Spelled out as a requirement: To persist state it must be exposed as outputs & inputs.